### PR TITLE
SW-3951 backend refactor the printer profile class and use it to implement rotary profile

### PR DESCRIPTION
**Pull request by khaledsherkawi**
SW-3951 Add profile class
The profile class is a modified version of the printerProfile class from OctoPrint. It has been modified to be able to serve as a parent class that can be implemented for different purposes and for different profile schemas.

SW-3951 Add laser cutter profile model

SW-3951 Add laser cutter profile service
This class implements the profile parent class to be used for laser cutter profiles.

SW-3951 Add laser cutter profiles to constants

SW-3951 Refactor services

SW-3951 Remove duplicated class

SW-3951 Refactor laser cutter profiles viewmodel

SW-3951 Implement the new laser cutter profile service

### CODE REVIEW

File: octoprint_mrbeam/__init__.py [hunk 1]
```diff
@@ -301,8 +296,23 @@ def initialize(self):
...
+    def update_laser_cutter_profile(self, profile):
+        self.laser_cutter_profile_service.save(profile, allow_overwrite=True, make_default=True)
+        self.laser_cutter_profile_service.select(profile['id'])
+
+    def get_laser_cutter_profile_for_current_configuration(self):
+        laser_cutter_mode = self.get_laser_cutter_mode()
+        device_series = self._device_info.get_series()
+        profile = laser_cutter_profiles.default_profile
+        if laser_cutter_mode == LaserCutterModeEnum.DEFAULT.value:
+            if device_series == "2C":
```
Comment (irlaec) [inline octoprint_mrbeam/__init__.py]: Do we have the device series somewhere in a constant? Or maybe it's a good moment to create an enum :slightly_smiling_face:

File: octoprint_mrbeam/__init__.py [hunk 2]
```diff
@@ -335,14 +345,29 @@ def _on_laserhead_ready(self, *args, **kwargs):
         self._laserhead_ready = True
         self._try_to_connect_laser()
 
+    def _on_laser_cutter_profile_initialized(self, *args, **kwargs):
```
Comment (irlaec) [inline octoprint_mrbeam/__init__.py:372]: If we don't need the *args, **kwargs, is it possible to remove them? like
```suggestion
    def _on_laser_cutter_profile_initialized(self, _, _):
```
Or maybe just remove it from the docstring

File: octoprint_mrbeam/__init__.py [hunk 3]
```diff
@@ -335,14 +345,29 @@ def _on_laserhead_ready(self, *args, **kwargs):
...
+        """Called when the laser cutter profile is initialized is initialized.
```
Comment (irlaec) [inline octoprint_mrbeam/__init__.py]:
```suggestion
        """Called when the laser cutter profile is initialized.
```

File: octoprint_mrbeam/__init__.py [hunk 4]
```diff
@@ -1700,31 +1725,16 @@ def engraveCalibrationMarkers(self, intensity, feedrate):
     # 		return NO_CONTENT
 
     # Laser cutter profiles
-    @octoprint.plugin.BlueprintPlugin.route("/profiles", methods=["GET"])
+    @octoprint.plugin.BlueprintPlugin.route("/currentProfile", methods=["GET"])
     def laserCutterProfilesList(self):
```
Comment (irlaec) [inline octoprint_mrbeam/__init__.py]: I think the name is for the old functionality. This is just a suggestion, just choose one following the convention if we have any:
```suggestion
    def get_current_laser_cutter_profile(self):
```

File: octoprint_mrbeam/__init__.py [hunk 5]
```diff
@@ -2117,7 +2127,7 @@ def on_api_command(self, command, data):
                         parse_csv(
                             device_model=self.get_model_id(),
                             laserhead_model=self.get_current_laser_head_model(),
-                            laser_cutter_mode=self.get_laser_cutter_mode(),
+                            laser_cutter_mode = self.get_laser_cutter_mode() if self.get_laser_cutter_mode() is not None else LaserCutterModeEnum.DEFAULT.value,
```
Comment (irlaec) [inline octoprint_mrbeam/__init__.py]: Suggestion (no need to change it if you prefer the other way):
```suggestion
                            laser_cutter_mode = self.get_laser_cutter_mode() or LaserCutterModeEnum.DEFAULT.value,
```

File: octoprint_mrbeam/constant/profile/laser_cutter/series_2c.py [hunk 6]
```diff
@@ -1,12 +1,17 @@
 #!/usr/bin/env python3
 
 profile = dict(
-    id="MrBeam2C",
+    id="series_2c",
```
Comment (irlaec) [inline octoprint_mrbeam/constant/profile/laser_cutter/series_2c.py]: Maybe we could use here the constant/enum I mentioned before

File: octoprint_mrbeam/constant/profile/laser_cutter/__init__.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+from __future__ import absolute_import
+
+from octoprint.util import dict_merge
+from . import default, series_2c, rotary, series_2c_rotary
```
Comment (irlaec) [inline octoprint_mrbeam/constant/profile/laser_cutter/__init__.py]: [PEP8](https://peps.python.org/pep-0008/#imports:~:text=Absolute%20imports%20are%20recommended%2C%20as%20they%20are%20usually%20more%20readable%20and%20tend%20to%20be%20better%20behaved%20(or%20at%20least%20give%20better%20error%20messages)%20if%20the%20import%20system%20is%20incorrectly%20configured%20(such%20as%20when%20a%20directory%20inside%20a%20package%20ends%20up%20on%20sys.path)%3A) recommends absolute imports

File: octoprint_mrbeam/printing/comm_acc2.py [hunk 8]
```diff
@@ -164,7 +164,7 @@ def __init__(
         self._port = port
         self._baudrate = baudrate
         self._callback = callbackObject
-        self._laserCutterProfile = laserCutterProfileManager().get_current_or_default()
+        self._laserCutterProfile = printerProfileManager.get_current_or_default()
```
Comment (irlaec) [inline octoprint_mrbeam/printing/comm_acc2.py]: In the __init__ the printerProfileManager argument is optional... should we change it? Otherwise we need to handle the case where it is None.
And why do we use the printerProfileManager now? We can discuss it on a call :slightly_smiling_face:

File: octoprint_mrbeam/printing/printer.py [hunk 9]
```diff
@@ -72,8 +74,8 @@ def connect(self, port=None, baudrate=None, profile=None):
         if self._comm is not None:
             self._comm.close()
 
-        eventManager().fire(Events.CONNECTING, payload=dict(profile=profile))
-        self._printerProfileManager.select(profile)
+        eventManager().fire(Events.CONNECTING, payload=dict(profile=self._printerProfileManager.get_current_or_default()['id']))
+        # self._printerProfileManager.select(profile)
```
Comment (irlaec) [inline octoprint_mrbeam/printing/printer.py:81]: Please remove this commented out code if we don't need it

File: octoprint_mrbeam/printing/printer.py [hunk 10]
```diff
@@ -48,6 +49,7 @@ def __init__(self, fileManager, analysisQueue, printerProfileManager):
             current_z=None,
         )
         self._user_notification_system = None
+        self._printerProfileManager = laser_cutter_profile_service()
```
Comment (irlaec) [inline octoprint_mrbeam/printing/printer.py:54]: Should it be _laser_cutter_profile_service instead, or is this something related to Octoprint? If it's the second, could you add a comment please?

File: octoprint_mrbeam/printing/printer.py [hunk 11]
```diff
@@ -112,7 +131,7 @@ def cancel_print(self):
         """Cancel the current printjob and do homing."""
         super(Laser, self).cancel_print()
         time.sleep(0.5)
-        self.home(axes="wtf")
```
Comment (irlaec) [inline octoprint_mrbeam/printing/printer.py:115]: :clap:

File: octoprint_mrbeam/printing/printer.py [hunk 12]
```diff
@@ -92,18 +94,35 @@ def set_colors(self, currentFileName, value):
...
+            after_homing_shift_rate=printer_profile["volume"]["after_homing_shift_rate"],
         )
         self._comm.rescue_from_home_pos()
         command = "G92X{x}Y{y}Z{z}".format(**params)
-        self.commands(["$H", command, "G90", "G21"])
+
+        # Moving command after homing
+        moving_command = "G1X{after_homing_shift_x}Y{after_homing_shift_y}F{after_homing_shift_rate}".format(**params)
+
+        """
```
Comment (irlaec) [inline octoprint_mrbeam/printing/printer.py]: Maybe move this to the docstring? And we can add there the comment about the axes arg too

File: octoprint_mrbeam/service/profile/laser_cutter_profile.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            dict(),
+        )
+
+    return _instance
+
+
+class LaserCutterProfileService(ProfileService):
+    """ Service class for laser cutter profile. """
+
+    DEFAULT_PROFILE_ID = "default"
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/laser_cutter_profile.py]: We could use LaserCutterModeEnum.DEFAULT.value here

File: octoprint_mrbeam/service/profile/profile.py [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+    load and save profiles.
+
+    A profile is a ``dict`` of a certain structure.
+
+    Args:
+        default (dict): The default profile to use if no other profile is selected.
+        folder (str): The folder to store the profiles in.
+    """
+
+    DEFAULT_PROFILE_ID = "_default"
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:51]: And here too (if we can use "default" instead of "_default")

File: octoprint_mrbeam/service/profile/profile.py [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+
+        default_overrides["id"] = self.DEFAULT_PROFILE_ID
+        self.save(default_overrides)
+
+        settings().set(["profiles", self._id, "defaultProfile"], None, force=True)
+        settings().save()
+
+        self._logger.info("Migrated default profile from settings to {}.profile: {!r}".format(self.DEFAULT_PROFILE_ID, default_overrides))
+
+    def _verify_default_available(self):
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:93]: Let's discuss this in the call too

File: octoprint_mrbeam/service/profile/profile.py [hunk 16]
```diff
@@ -0,0 +1,482 @@
...
+            default_id = self.DEFAULT_PROFILE_ID
+
+        profile = self.get(default_id)
+        if profile is None:
+            self._logger.error("Selected default profile {} is invalid, resetting to default values".format(default_id))
+            profile = copy.deepcopy(self._default)
+            profile["id"] = default_id
+            self.save(self._default, allow_overwrite=True, make_default=True)
+
+    def select(self, identifier):
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:122]: For select, remove, etc: what is the boolean we return? Maybe we can make the names of the methods more self-explanatory, or at least add some docstrings

File: octoprint_mrbeam/service/profile/profile.py [hunk 17]
```diff
@@ -0,0 +1,482 @@
...
+    def save(self, profile, allow_overwrite=False, make_default=False):
+        if "id" in profile:
+            identifier = profile["id"]
+        elif "name" in profile:
+            identifier = profile["name"]
+        else:
+            raise InvalidProfileError("profile must contain either id or name")
+
+        identifier = self._sanitize(identifier)
+        profile["id"] = identifier
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:164]: Maybe up to this point we could have a single method called something like _generate_profile_id(profile)

Comment (irlaec) [reply octoprint_mrbeam/service/profile/profile.py:164]: Or continue more, and just _sanitize_profile(profile) where we do everything until _save_to_path

File: octoprint_mrbeam/service/profile/profile.py [hunk 18]
```diff
@@ -0,0 +1,482 @@
...
+
+    @property
+    def last_modified(self):
+        dates = [os.stat(self._folder).st_mtime]
+        dates += [entry.stat().st_mtime for entry in scandir(self._folder) if entry.name.endswith(".profile")]
+        return max(dates)
+
+    def get_default(self):
+        default = settings().get(["profiles", self._id, "default"])
+        if default is not None and self.exists(default):
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:196]: I think this could be simplified, as we already handle the None value inside of self.exists()
```suggestion
        if self.exists(default):
```

Comment (irlaec) [reply octoprint_mrbeam/service/profile/profile.py:196]: Or even more, in self.get() we already check if it exists, so we wouldn't need this statement

File: octoprint_mrbeam/service/profile/profile.py [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+            if profile is not None:
+                return profile
+            else:
+                self._logger.warn("Default profile {} is invalid, falling back to built-in defaults".format(default))
+
+        return copy.deepcopy(self._default)
+
+    def set_default(self, identifier):
+        all_identifiers = self._load_all_identifiers().keys()
+        if identifier is not None and not identifier in all_identifiers:
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py]: I think it can be simplified:
```suggestion
        if identifier not in all_identifiers:
```

File: octoprint_mrbeam/service/profile/profile.py [hunk 20]
```diff
@@ -0,0 +1,482 @@
...
+
+            identifier = entry.name[:-len(".profile")]
+            results[identifier] = entry.path
+        return results
+
+    def _load_from_path(self, path):
+        if not os.path.exists(path) or not os.path.isfile(path):
+            return None
+
+        import yaml
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:262]: Can you please move this import to the top?

File: octoprint_mrbeam/service/profile/profile.py [hunk 21]
```diff
@@ -0,0 +1,482 @@
...
+        with open(path) as f:
+            profile = yaml.safe_load(f)
+
+        if profile is None or not isinstance(profile, dict):
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:266]:
```suggestion
        if not isinstance(profile, dict):

```

File: octoprint_mrbeam/service/profile/profile.py [hunk 22]
```diff
@@ -0,0 +1,482 @@
...
+
+    def _save_to_path(self, path, profile, allow_overwrite=False):
+        validated_profile = self._ensure_valid_profile(profile)
+        if not validated_profile:
+            raise InvalidProfileError()
+
+        if os.path.exists(path) and not allow_overwrite:
+            raise SaveError("Profile %s already exists and not allowed to overwrite" % profile["id"])
+
+        from octoprint.util import atomic_write
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py:292]: Same with these imports

File: octoprint_mrbeam/service/profile/profile.py [hunk 23]
```diff
@@ -0,0 +1,482 @@
...
+        sanitized_name = ''.join(c for c in name if c in valid_chars)
+        sanitized_name = sanitized_name.replace(" ", "_")
+        return sanitized_name
+
+    def _migrate_profile(self, profile):
+        """
+        Subclasses must override this method.
+        """
+        return False
+        # raise NotImplementedError("Subclasses must implement _migrate_profile.")
```
Comment (irlaec) [inline octoprint_mrbeam/service/profile/profile.py]: In the call, let's discuss
- I would maybe make the names of the methods in the profile service more self-explanatory or add docstrings
- What exactly is "identifier", because it's different to the id, no?
- What do we need this commented out code for?

**Review by irlaec — CHANGES_REQUESTED** [review]

File: octoprint_mrbeam/__init__.py (hunk 1 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: The list of device series was removed from the plugin implementation because it leads to unneeded overhead every time we add a new series. Now the series comes from the /etc/mrbeam file always.

Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: I will add the Enum for the relevant device series though! Great point

File: octoprint_mrbeam/__init__.py (hunk 2 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py:372]: This is fine as we can add any relevant arguments to this later. This is similar to other event handlers we are using elsewhere.

Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py:372]: I will remove from docstring though

File: octoprint_mrbeam/__init__.py (hunk 4 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: On point. I will implement it.

File: octoprint_mrbeam/__init__.py (hunk 3 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: I will implement it.

File: octoprint_mrbeam/__init__.py (hunk 5 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: I will implement it.

File: octoprint_mrbeam/constant/profile/laser_cutter/__init__.py (hunk 7 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/constant/profile/laser_cutter/__init__.py]: I will implement.

File: octoprint_mrbeam/constant/profile/laser_cutter/series_2c.py (hunk 6 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/constant/profile/laser_cutter/series_2c.py]: I will implement.

File: octoprint_mrbeam/printing/printer.py (hunk 10 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/printing/printer.py:54]: Yes, this is an attribute we are overriding from OctoPrint. I will add a comment

File: octoprint_mrbeam/printing/printer.py (hunk 9 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/printing/printer.py:81]: No this is left on purpose for later when we want to remove the restart/reload logic

File: octoprint_mrbeam/printing/printer.py (hunk 12 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/printing/printer.py]: I will implement this.

File: octoprint_mrbeam/service/profile/laser_cutter_profile.py (hunk 13 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/laser_cutter_profile.py]: I will implement this.

File: octoprint_mrbeam/service/profile/profile.py (hunk 14 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:51]: This is basically the parent class that will always be implemented to be used and not directly used. So the values here are just for demonstration

File: octoprint_mrbeam/service/profile/profile.py (hunk 15 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:93]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 16 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:122]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 17 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:164]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 18 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:196]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 19 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 20 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:262]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 21 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:266]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 22 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py:292]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 23 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py]: This is not my implementation but OctoPrint's. We are using their class as a parent class with minor changes to fir our purpose

File: octoprint_mrbeam/service/profile/profile.py (hunk 19 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/service/profile/profile.py]: @irlaec why is this left unresolved?

Comment (irlaec) [reply octoprint_mrbeam/service/profile/profile.py]: I didn't notice it 😅

**Review by irlaec — APPROVED** [review]
